File: source/com/barrybecker4/puzzle/common/ui/PuzzleApplet.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.puzzle.common.ui;

import com.barrybecker4.puzzle.common.AlgorithmEnum;
import com.barrybecker4.puzzle.common.PuzzleController;
import com.barrybecker4.common.search.Refreshable;
import com.barrybecker4.ui.application.ApplicationApplet;

import javax.swing.JPanel;
import java.awt.BorderLayout;
import java.util.ArrayList;
import java.util.List;

/**
 * Base class for Puzzle applets.
 *
 * @author Barry Becker
 */
public abstract class PuzzleApplet<P, M> extends ApplicationApplet {

    protected PuzzleController<P, M> controller_;
    protected PuzzleViewer<P, M> viewer_;

    private static final String DEFAULT_PUZZLE = "com.barrybecker4.puzzle.maze.FractalExplorer";


    /**
     * Construct the applet.
     * No argument constructor needed for applet instantiation by reflection.
     */
    public PuzzleApplet()  {
        super(new String[] {});
    }

    /**
     * Construct the application.
     */
    public PuzzleApplet(String[] args) {
        super(args);
    }

    /**
     * create and initialize the puzzle
     * (init required for applet)
     */
    @Override
    protected JPanel createMainPanel() {
        JPanel mainPanel = new JPanel(new BorderLayout());

        viewer_ = createViewer();
        controller_ = createController(viewer_);
        viewer_.refresh(controller_.initialState(), 0);

        TopControlPanel<P, M> topPanel = createTopControls();

        mainPanel.add(topPanel, BorderLayout.NORTH);
        mainPanel.add(viewer_, BorderLayout.CENTER);
        JPanel bottomControls = createBottomControls();
        if (bottomControls != null) {
            mainPanel.add(bottomControls, BorderLayout.SOUTH);
        }
        return mainPanel;
    }

    protected List<String> getResourceList() {
        List<String> resources = new ArrayList<>(super.getResourceList());
        resources.add("com.barrybecker4.puzzle.common.ui.message");
        return resources;
    }

    protected abstract PuzzleViewer<P, M> createViewer();

    protected abstract PuzzleController<P, M> createController(Refreshable<P, M> viewer);

    protected TopControlPanel<P, M> createTopControls() {
        return new TopControlPanel<>(controller_, getAlgorithmValues());
    }

    protected JPanel createBottomControls() {
        return null;
    }

    protected abstract AlgorithmEnum<P, M>[] getAlgorithmValues();

}

File: scala-source/com/barrybecker4/puzzle/common/ui/PuzzleApplet.scala
// Copyright by Barry G. Becker, 2000-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.common.ui

import com.barrybecker4.puzzle.common.AlgorithmEnum
import com.barrybecker4.puzzle.common.PuzzleController
import com.barrybecker4.search.Refreshable
import com.barrybecker4.ui.application.ApplicationApplet
import javax.swing.JPanel
import java.awt.BorderLayout
import java.util


/**
  * Base class for Puzzle applets.
  *
  * @author Barry Becker
  */
object PuzzleApplet {
  private val DEFAULT_PUZZLE = "com.barrybecker4.puzzle.maze.FractalExplorer"
}

abstract class PuzzleApplet[P, M](args: Array[String])  extends ApplicationApplet() {
  protected var controller:
  PuzzleController[P, M] = _
  protected var viewer: PuzzleViewer[P, M] = _

  /** Construct the application. */
  def this() { this(Array[String]())}

  /**
    * Create and initialize the puzzle
    * (init required for applet)
    */
  override protected def createMainPanel: JPanel = {
    val mainPanel = new JPanel(new BorderLayout)
    viewer = createViewer
    controller = createController(viewer)
    viewer.refresh(controller.initialState, 0)
    val topPanel = createTopControls
    mainPanel.add(topPanel, BorderLayout.NORTH)
    mainPanel.add(viewer, BorderLayout.CENTER)
    val bottomControls = createBottomControls
    if (bottomControls != null)
      mainPanel.add(bottomControls, BorderLayout.SOUTH)
    mainPanel
  }

  override protected def getResourceList: util.List[String] = {
    val resources = new util.ArrayList[String](super.getResourceList)
    resources.add("com.barrybecker4.puzzle.common.ui.message")
    resources
  }

  protected def createViewer: PuzzleViewer[P, M]

  protected def createController(viewer: Refreshable[P, M]): PuzzleController[P, M]

  protected def createTopControls: JPanel = new TopControlPanel[P, M](controller, getAlgorithmValues)

  protected def createBottomControls: JPanel

  protected def getAlgorithmValues: Array[AlgorithmEnum[P, M]]
}

